addons/hr_holidays/models/hr_leave.py

```diff
@@ -266,6 +266,17 @@ def _compute_request_hour_from_to(self):
                 leave.request_hour_from = hour_from
                 leave.request_hour_to = hour_to
 
+    @api.onchange('request_date_from', 'request_date_to')
+    def _onchange_dates(self):
+        if not self.holiday_status_id.get_leave_types_with_valid_allocations(self.request_date_from, self.request_date_to, self.employee_id.id):
+            valid_types = self.env['hr.leave.type'].search([]).get_leave_types_with_valid_allocations(self.request_date_from, self.request_date_to, self.employee_id.id)
+            if not valid_types or not self.employee_id:
+                domain = [('requires_allocation', '=', False)]
+                leave_types = self.env['hr.leave.type'].search(domain)
+                self.holiday_status_id = leave_types[0] if leave_types else None
+            elif valid_types:
+                self.holiday_status_id = valid_types[0]
```

Comment:
looks like it should be in the compute `_compute_from_employee_id` which could take the request dates as an extra dependency (which might be renamed as `_compute_holiday_status_id`)